cloudinstall/core.py

```diff
@@ -534,14 +533,35 @@ def get_machine_spec(self, maas_machine, atype):
             log.error("unexpected atype: {}".format(atype))
             return None
 
+    def wait_for_deployed_services_ready(self):
+        """ Blocks until all deployed services attached units
+        are in a 'started' state
+        """
+        self.ui.status_info_message(
+            "Waiting for deployed services to be in a ready state.")
+
+        not_ready_len = 0
+        while not self.juju_state.all_agents_started():
+            not_ready = [(a, b) for a, b in self.juju_state.get_agents_states()
+                         if b != 'started']
+            if len(not_ready) == not_ready_len:
+                time.sleep(3)
+                continue
+
+            not_ready_len = len(not_ready)
+            log.info("Checking availability of {} ".format(
+                ", ".join(["{}:{}".format(a, b) for a, b in not_ready])))
+            time.sleep(3)
+
+        self.ui.status_info_message(
+            "Processing Relations and finalizing services")
+
     def enqueue_deployed_charms(self):
         """Send all deployed charms to CharmQueue for relation setting and
         post-proc.
         """
 
-        self.ui.status_info_message(
-            "Processing Relations and finalizing charm operations"
-            "{}".format([c.charm_name for c in self.deployed_charm_classes]))
+        self.wait_for_deployed_services_ready()
```

Comment:
I think this line should be moved to `begin_deployment()`, between lines 279 and 280. That'll make it more obvious what's going on at that level, and `enqueue_deployed_charms` will only do what it says it does and no more.